sdk/nodejs/log/index.ts

```diff
@@ -118,5 +125,13 @@ function log(
         });
     });
 
-    return lastLog.catch(() => undefined);
+    return lastLog.catch((err) => {
+        // debug messages never go to stdout/err
+        if (sev !== engproto.LogSeverity.DEBUG) {
+            // if we're unable to deliver the log message, deliver to stderr instead
+            console.error(`failed to deliver log message. \nerror: ${err} \noriginal message: ${msg}\n message severity: ${messageLevels[sev]}`)
```

Comment:
When will this happen?  I expect ideally we don't ever want to show this - but if it's only shown when there is truly a bug in Pulumi of some kind, I suppose this is okay as an assert type of condition.

Reply:
A few examples of failure modes:
- For a log with an attached resource, the URN promise rejects (the resource fails to create)
- Some cases where resource monitor requests fail, we shut down all in flight RPC requests, even logging events sent to the engine. This is what is outlined in #6712 https://github.com/pulumi/pulumi/blob/d098f9181b114fd97b0d4caba3d883a5ca8b0190/sdk/nodejs/runtime/resource.ts#L721

Reply:
Also, just consider the general case that this RPC fails. The users then loses the log. This makes a best-effort attempt to deliver in some form.

Reply:
Got it. 

My concern is that this should be very rare. I very much hope that users can defectively never see this - but slightly worry that (especially due to #6712) we will start seeing this error a lot in "normal" cases. 

> For a log with an attached resource, the URN promise rejects (the resource fails to create)

I will say though, at least in this case, if not in all cases, it seems it would be better to bubble this up as an unhandled rejection than to try and `console.error` it. This feels like it's circumventing proper error handling, and providing another awkward path out of bubbling this up normally. 

I suppose the argument perhaps is that although `log` returns a promise, in practice noone actually listens to and handles that promise, and we don't want to tell them to, so we have to handle the failure somehow ourselves instead of bubbling it up "correctly"?

Reply:
If we do not handle the error, then the user gets unhandled rejection output ala:

```js
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 2)
    (Use `node --trace-warnings ...` to show where the warning was created)
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 3)
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 4)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 5)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 6)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 7)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 8)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 9)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 10)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 11)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 12)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 13)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 14)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 15)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 16)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 17)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 18)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 19)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 21)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 23)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 25)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 27)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 29)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 31)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 32)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 33)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 34)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 35)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 36)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 37)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 38)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 39)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 40)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 41)
    Error: foo abc
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:104:23
        at new Promise (<anonymous>)
        at /opt/pulumi/node_modules/@pulumi/pulumi/log/index.js:96:16
    (node:25844) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 42)

```

And the program still succeeds. 

My change alters this output to instead look like:

```
    failed to deliver log message.
    error: Error: foo
    original message: foo foo
     message severity: warn
    failed to deliver log message.
    error: Error: foo
    original message: foo foo
     message severity: warn
    failed to deliver log message.
    error: Error: foo
    original message: foo foo
     message severity: warn
    failed to deliver log message.
    error: Error: foo
    original message: foo foo
     message severity: warn
    failed to deliver log message.
    error: Error: foo
    original message: foo foo
     message severity: warn
    failed to deliver log message.
    error: Error: foo
    original message: foo foo
     message severity: warn

```

Reply:
They didn't used to get these errors until recently - right?  What exactly changed that causes the errors now?

There was already code here that catches and stops rejections from being unhandled. I'm having trouble understanding why this change "fixes" anything. Is it the keep alive call?

Did we used to see these log messages just silently not logged?  Is that why we newly need a console.error here?

End of the day - happy to take this change - but I don't quite understand why it's "right", and it feels like there is some deeper issues here.